### ISSUE TITLE
Proposal: Extract method to new class

### ISSUE THREAD

**Issue by Zyxious** [JosefPihrt/Roslynator#221]
I want to take a existing method in a class and move this to a new class. This is useful while doing refactoring.

By placing the cursor on a method, I would  like to get a choice like:
'Move method to new class'

This would for example move the method to a new class in the same file and let me edit the class name. After that I can us a existing refactoring to move the class to a new file. 

Alternatively we could have another choice like:
'Move method to new file and class'

This would move the method to a new class in a new file. Maybe there is a little bit of problem naming the file since the class name is not defined.

**Comment by wmjordan** [comment]
I vote for this request.
While I was reviewing the code written by my team member, I sometimes gave them suggestions to move some methods to a separated class, for the reason that a class should be kept as simple as doing one thing.
I myself occasionally did such type of refactoring as well.

**Comment by GrahamTheCoder** [comment]
This is one of ReSharper's most valuable refactorings IMO. Their implementation is somewhat buggy, but consistently so, and it gets close enough to still be of great value. The main good thing about their feature is that it lets you choose multiple things to move at the same time (see [their docs](https://www.jetbrains.com/resharper/features/screenshots/20161/refactoring_extract_class.png) if unfamiliar), and lets you decide what to do about fields that are reused etc (though usually the answer is you want to create a copy in both and pass them via the constructor if they're used in both so the choice isn't as important as visualising dependencies I think)

**Comment by wmjordan** [comment]
@GrahamTheCoder You are right. Even though there may be bugs, it is still valuable. I think I can accept the buggy one.

**Comment by kepiz** [comment]
Will this be implemented, I think it is one of the most useful features refactoring, and cant do it in VS code

**Comment by marinasundstrom** [comment]
This seems like a fun weekend project!

The issue is finding out what members the method is referencing and see whether that prevents it from being extracted.

But it is kind of useless if you can't pick multiple members to extract.

If there is dependency injection involved you have to figure those dependencies out too for the target class.

Safe to say, this feature won't be 100% perfect, but at least it can guide you.

**Comment by GrahamTheCoder** [comment]
Yeah if you pull out something dependent on a field, it'd be nice to inject a copy of that field to the new class (could warn about non-readonly fields).
If it depends on a private/protected method, usually it should also be extracted and made public if needed.

The ReSharper implementation injects "this" into the constructor of the new class and lets everything reference back via that, introducing a cyclic dependency which I basically never want.

**Comment by marinasundstrom** [comment]
@GrahamTheCoder I have asked ChatGPT. It gave me some code. Let's see if I get around testing it.
